### PR TITLE
chore: replace black with ruff as the formatter

### PR DESCRIPTION
**Pull request by wood-push-melon**
This pull request:

- replaces black with ruff as the formatter
- fixes the lint and format issues with ruff
- keeps the `isort` to keep the original behaviors for import sorting
- puts the `codespell` configuration to the `pyproject.toml` file to be consistent with other tools

### CODE REVIEW

File: lint-requirements.txt [hunk 1]
```diff
@@ -1,9 +1,2 @@
-black
-flake8==6.1.0
-flake8-docstrings
-flake8-copyright
```
Comment (natalian98) [inline lint-requirements.txt:4]: Does ruff check for copyright notices? I think we'd need to use `--preview` flag: https://docs.astral.sh/ruff/rules/missing-copyright-notice/

File: pyproject.toml [hunk 2]
```diff
@@ -12,33 +12,50 @@ show_missing = true
...
+    "E501",
+    "N818"
+]
+per-file-ignores = {"tests/*" = ["D100","D101","D102","D103","D104"]}
 
-# Linting tools configuration
-[tool.flake8]
-max-line-length = 99
-max-doc-length = 99
+[too.ruff.lint.flake8-copyright]
```
Comment (natalian98) [inline pyproject.toml:50]:
```suggestion
[tool.ruff.lint.flake8-copyright]
```

File: pyproject.toml [hunk 3]
```diff
@@ -12,33 +12,50 @@ show_missing = true
...
 line-length = 99
-target-version = ["py38"]
+include = ["pyproject.toml", "src/**/*.py", "tests/**/*.py", "lib/charms/hydra/**/*.py"]
+extend-exclude = ["__pycache__", "*.egg_info"]
+target-version = "py38"
 
-[tool.isort]
-line_length = 99
-profile = "black"
+[too.ruff.lint]
```
Comment (natalian98) [inline pyproject.toml:24]:
```suggestion
[tool.ruff.lint]
```

File: lint-requirements.txt (hunk 1 above)
Comment (wood-push-melon) [reply lint-requirements.txt:4]: It does when `preview` is enabled. Just somehow it keeps complaining about it so I disabled it for now. I am currently taking a look at the `CPY` rule and checking where goes wrong.

File: src/charm.py [hunk 4]
```diff
@@ -58,15 +59,12 @@
...
-    ModelError,
     Relation,
     Secret,
     SecretNotFoundError,
     WaitingStatus,
 )
 from ops.pebble import ChangeError, Error, ExecError, Layer
-
-from hydra_cli import HydraCLI
 from utils import normalise_url, remove_none_values
```
Comment (nsklikas) [inline src/charm.py:68]: Import sorting seems to have changed, I think we should keep the current approach where local imports are separately grouped

File: tox.ini [hunk 5]
```diff
@@ -28,30 +28,27 @@ deps =
...
+    ruff format {[vars]all_path}
 
 [testenv:lint]
 description = Check code against coding style standards
 deps =
+    tomli
     -r{toxinidir}/lint-requirements.txt
 commands =
-    codespell {[vars]lib_path}
-    codespell {toxinidir}/ --skip {toxinidir}/.git --skip {toxinidir}/.tox \
```
Comment (wood-push-melon) [inline tox.ini:48]: Moved the configuration of `codespell` to `pyproject.toml` file.

File: tox.ini [hunk 6]
```diff
@@ -28,30 +28,27 @@ deps =
...
 
 [testenv:fmt]
-description = Apply coding style standards to code
+description = Apply coding style standards
 deps =
     -r{toxinidir}/fmt-requirements.txt
 commands =
-    isort {[vars]all_path}
-    black {[vars]all_path}
+    ruff format {[vars]all_path}
```
Comment (nsklikas) [inline tox.ini:41]: Don't we need to run isort here as well?

**Review by nsklikas — APPROVED** [review]
lgtm

**Pull request merged**